subsys/net/lib/nrf_cloud/src/nrf_cloud_net_utils.c

```diff
@@ -0,0 +1,144 @@
+
+	err = getaddrinfo(host_name, NULL, hints, &info);
+	if (err) {
+		LOG_DBG("getaddrinfo for %s, port: %d failed: %d, errno: %d",
+			host_name, port, err, errno);
+		return -EADDRNOTAVAIL;
+	}
+
+	/* Hold copy of original result so that we can free it later. */
+	result = info;
+
+	/* Try to connect with whatever IP addresses we get.
+	 * Stop and return the socket if one succeeds.
+	 */
+	for (; info != NULL; info = info->ai_next) {
+		char ip[INET6_ADDRSTRLEN] = { 0 };
+		struct sockaddr *const sa = info->ai_addr;
+
+		switch (sa->sa_family) {
+		case AF_INET6:
+			((struct sockaddr_in6 *)sa)->sin6_port = port;
+			break;
+		case AF_INET:
+			((struct sockaddr_in *)sa)->sin_port = port;
+			break;
+		}
+
+		inet_ntop(sa->sa_family, (void *)&((struct sockaddr_in *)sa)->sin_addr,
+			  ip, sizeof(ip));
+
+		LOG_DBG("Trying IP address and port for server %s: %s, port: %d",
+			host_name, ip, port);
+
+		/* The try_socket callback will attempt to create a socket, and either return that,
+		 * or a negative error code.
+		 */
+		int sock = connect_cb(sa);
+
+		if (sock < 0) {
+			LOG_DBG("Failed to connect to IP address %s. Error: %d", ip, sock);
+			continue;
+		}
+
+		LOG_INF("Chosen IP address for server %s is %s", host_name, ip);
+
+		/* Pass the socket back to the initial caller, if creating/connecting it was
+		 * successful.
+		 */
+		freeaddrinfo(result);
+		return sock;
+	}
+
+	freeaddrinfo(result);
+	return -ECONNREFUSED;
+}
+
+int nrf_cloud_connect_host(const char *hostname, uint16_t port, struct addrinfo *hints,
+			   nrf_cloud_try_address_cb connect_cb)
+{
+	/* Query IPv6 first, then IPv4. This is necessary because some socket backends do not
+	 * return all available IP addresses, so if we just use AF_UNSPEC, we might get only an
+	 * IPv6 address, even if both IPv4 and IPv6 are available.
+	 */
+	int sock;
+
+	if (hints == NULL) {
+		return -EINVAL;
+	}
+
+	LOG_DBG("Connecting to nRF Cloud");
+
+#if defined(CONFIG_NRF_CLOUD_STATIC_IPV4) && defined(CONFIG_NET_IPV4)
+	static struct sockaddr_in static_addr;
+
+	LOG_DBG("Trying static IPv4 address: %s", CONFIG_NRF_CLOUD_STATIC_IPV4_ADDR);
+
+	inet_pton(AF_INET, CONFIG_NRF_CLOUD_STATIC_IPV4_ADDR, &static_addr);
+	static_addr->sin_family = AF_INET;
+	static_addr->sin_port = htons(NRF_CLOUD_PORT);
+
+	int sock = connect_cb(sock);
+
+	if (sock >= 0) {
+		goto out;
+	}
+
+	LOG_DBG("Failed to connect to IP address %s", ip);
+#endif
+
+#if defined(CONFIG_NET_IPV6)
+	LOG_DBG("Trying IPv6 addresses for %s", hostname);
+
+	hints->ai_family = AF_INET6;
+	sock = nrf_cloud_try_addresses(hostname, port, hints, connect_cb);
+
+	if (sock >= 0) {
+		goto out;
+	}
+
+#endif /* defined (CONFIG_NET_IPV6) */
+
+#if defined(CONFIG_NET_IPV4) && !defined(CONFIG_NRF_CLOUD_IPV6)
+	LOG_DBG("Trying IPv4 addresses for %s", hostname);
+
+	hints->ai_family = AF_INET;
+	sock = nrf_cloud_try_addresses(hostname, port, hints, connect_cb);
+
+	if (sock >= 0) {
+		goto out;
+	}
+
+#endif /* defined (CONFIG_NET_IPV4) */
```

Comment:
is it possible to successfully build without `CONFIG_NET_IPV6` or `CONFIG_NET_IPV4` enabled?
if it is, add a build assert?
if not, ignore.

Reply:
This is one of the things I still need to test, but in theory it should work fine, since the headers/types are always exposed. This kind of IPV4/IPV6 modularity is commonly done in Zephyr

Reply:
>  in theory it should work fine

if either is not defined, `nrf_cloud_try_addresses` won't be called.

Reply:
Aha, I will check that too, good point. I think at least one has to be enabled, but I'll find out for sure.